coreos-ostree-importer/coreos_ostree_importer.py

```diff
@@ -0,0 +1,280 @@
+def get_sha256sum(filepath: str) -> str:
+    h = hashlib.sha256()
+    with open(filepath, "rb") as file:
+        while True:
+            # Reading is buffered, so we can read smaller chunks.
+            chunk = file.read(h.block_size)
+            if not chunk:
+                break
+            h.update(chunk)
+    return h.hexdigest()
+
+
+def untar_file_from_s3(url: str, tmpdir: str, sha256sum: str):
+    filename = "ostree.tar"
+    filepath = os.path.join("/tmp/", filename)
+
+    # Grab file from s3
+    logger.info(f"Downloading object from s3: {url}")
+    s3 = boto3.client("s3")
+    bucket, key = parse_s3_url(url)
+    s3.download_file(bucket, key, filepath)
+
+    # Verify file has correct checksum
+    calcuatedsum = get_sha256sum(filepath)
+    if sha256sum != calcuatedsum:
+        raise Exception("Checksums do not match: " f"{sha256sum} != {calcuatedsum}")
+
+    # Untar the file into the temporary directory
+    with tarfile.open(filepath) as tar:
+        tar.extractall(path=tmpdir)
+
+
+def commit_exists(repo: str, commit: str) -> bool:
+    if not ostree_repo_exists(repo):
+        return False
+    else:
+        cmd = ["ostree", f"--repo={repo}", "show", commit]
+        return runcmd(cmd, check=False).returncode == 0
+
+
+def assert_commit_has_version(repo: str, commit: str, version: str):
+    cmd = ["ostree", f"--repo={repo}", "show", commit, "--print-metadata-key=version"]
+    cp = runcmd(cmd, check=True)
+    embeddedversion = cp.stdout.replace(b"'", b"").strip().decode()
+    if version != embeddedversion:
+        raise Exception(
+            "Embedded commit version does not match buildid "
+            f"{version} != {embeddedversion}"
+        )
+
+
+def ostree_repo_exists(repo: str) -> bool:
+    if not os.path.exists(repo):
+        return False
+    cmd = ["ostree", f"--repo={repo}", "refs"]
+    if runcmd(cmd, check=False).returncode != 0:
+        logger.debug(f"OSTree repo does not exist at {repo}")
+        return False
+    return True
+
+
+def runcmd(cmd: list, **kwargs: int) -> subprocess.CompletedProcess:
+    try:
+        # default args to pass to subprocess.run
+        pargs = {"check": True, "capture_output": True}
+        logger.debug(f"Running command: {cmd}")
+        pargs.update(kwargs)
+        cp = subprocess.run(cmd, **pargs)
+    except subprocess.CalledProcessError as e:
+        logger.error("Command returned bad exitcode")
+        logger.error(f"COMMAND: {cmd}")
+        logger.error(f" STDOUT: {e.stdout.decode()}")
+        logger.error(f" STDERR: {e.stderr.decode()}")
+        raise e
+    return cp  # subprocess.CompletedProcess
+
+
+def ostree_pull_local(srcrepo: str, dstrepo: str, branch: str, commit: str):
+    # pull content
+    cmd = ["ostree", f"--repo={dstrepo}", "pull-local", srcrepo, commit]
+    runcmd(cmd)
+    # update branch
+    if ostree_branch_exists(repo=dstrepo, branch=branch):
+        cmd = ["ostree", f"--repo={dstrepo}", "reset", branch, commit]
```

Comment:
See sanity-check related to this up above

Reply:
+1